pointmatcher/DataPointsFilters/Geometry.cpp

```diff
@@ -0,0 +1,170 @@
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+#include "Geometry.h"
+
+// Eigenvalues
+#include "Eigen/QR"
+#include "Eigen/Eigenvalues"
+
+#include "PointMatcherPrivate.h"
+#include "IO.h"
+#include "MatchersImpl.h"
+
+#include <boost/format.hpp>
+
+
+#include "utils.h"
+
+// SurfaceNormalDataPointsFilter
+// Constructor
+template<typename T>
+GeometryDataPointsFilter<T>::GeometryDataPointsFilter(const Parameters& params):
+	PointMatcher<T>::DataPointsFilter("GeometryDataPointsFilter",
+	        GeometryDataPointsFilter::availableParameters(), params),
+    keepUnstructureness(Parametrizable::get<int>("keepUnstructureness")),
+    keepStructureness(Parametrizable::get<T>("keepStructureness"))
+{
+}
+
+// Compute
+template<typename T>
+typename PointMatcher<T>::DataPoints
+GeometryDataPointsFilter<T>::filter(
+	const DataPoints& input)
+{
+	DataPoints output(input);
+	inPlaceFilter(output);
+	return output;
+}
+
+// In-place filter
+template<typename T>
+void GeometryDataPointsFilter<T>::inPlaceFilter(
+	DataPoints& cloud)
+{
+	typedef typename DataPoints::View View;
+	typedef typename DataPoints::Label Label;
+	typedef typename DataPoints::Labels Labels;
+
+	const int pointsCount(cloud.features.cols());
+	const int descDim(cloud.descriptors.rows());
+	const unsigned int labelDim(cloud.descriptorLabels.size());
+
+	// Check that the required eigenValue descriptor exists in the pointcloud
+    if (!cloud.descriptorExists("eigValues"))
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, no eigValues found in descriptors.");
+    }
+
+    // Validate descriptors and labels
+	int insertDim(0);
+	for(unsigned int i = 0; i < labelDim ; ++i)
+		insertDim += cloud.descriptorLabels[i].span;
+	if (insertDim != descDim)
+		throw InvalidField("SurfaceNormalDataPointsFilter: Error, descriptor labels do not match descriptor data");
+
+	// Reserve memory for new descriptors
+	const int dimSphericality(1);
+	const int dimUnstructureness(1);
+	const int dimStructureness(1);
+
+	boost::optional<View> sphericality;
+	boost::optional<View> unstructureness;
+	boost::optional<View> structureness;
+
+	Labels cloudLabels;
+	cloudLabels.push_back(Label("sphericality", dimSphericality));
+	if (keepUnstructureness)
+		cloudLabels.push_back(Label("unstructureness", dimUnstructureness));
+	if (keepStructureness)
+		cloudLabels.push_back(Label("structureness", dimStructureness));
+
+    // Reserve memory
+	cloud.allocateDescriptors(cloudLabels);
+
+	// Get the views
+    const View eigValues = cloud.getDescriptorViewByName("eigValues");
+    if (eigValues.rows() != 3)  // And check the dimensions
```

Comment:
From these lines I can infer that this filter works only in 3D, right?

Maybe the name of the filter class should reflect that. Or at least have a parameter to specify the mode?

Reply:
I think I'll rename the filter to SphericalLevelDataPointsFilter, which kinda indicates 3D.